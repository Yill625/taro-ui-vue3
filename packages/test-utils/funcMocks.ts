import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks