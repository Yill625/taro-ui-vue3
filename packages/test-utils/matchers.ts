import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers