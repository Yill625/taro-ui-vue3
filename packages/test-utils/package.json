{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.2.0",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}