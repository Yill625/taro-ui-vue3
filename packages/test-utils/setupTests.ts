import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()