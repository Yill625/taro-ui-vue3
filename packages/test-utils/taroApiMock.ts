import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
