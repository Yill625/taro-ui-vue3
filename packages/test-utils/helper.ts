import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}