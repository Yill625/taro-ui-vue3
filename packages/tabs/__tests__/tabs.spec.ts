import { mountFactory, Slots, triggerTouchEvents } from '@taro-ui-vue3/test-utils/helper'
import { h } from '@vue/runtime-core'
import { ref } from 'vue'
import AtTabs from '../index'

const factory = (props = {}, slots?: Slots) => {
  return mountFactory(AtTabs, undefined, props, slots)
}

const tabList = [
  { title: '标签页1' },
  { title: '标签页2' },
  { title: '标签页3' }
]

describe('AtTabs', () => {

  beforeEach(() => {
    jest.useFakeTimers()
    jest.advanceTimersByTime(100)
    jest.runAllTimers()
  })

  it('should render default AtTabs', () => {
    const wrapper = factory()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop -- style', () => {
    const wrapper = factory({ style: 'color: red;' })
    expect(
      wrapper
        .find('.at-tabs')
        .attributes('style')
    ).toContain('color: red;')
  })

  it('should render prop -- class', () => {
    const wrapper = factory({ class: 'test' })
    expect(
      wrapper
        .find('.at-tabs')
        .classes()
    ).toContain('test')
  })

  it('should render prop -- tabList', () => {
    const wrapper = factory({ tabList: tabList })
    expect(
      wrapper
        .findAll('.at-tabs__item')
        .length
    ).toBe(tabList.length)

    expect(
      wrapper
        .findAll('.at-tabs__item--active')
        .length
    ).toBe(1)

    expect(
      wrapper
        .find('.at-tabs__header')
        .element
    ).toMatchSnapshot()
  })

  it('should render prop -- current', async () => {
    const wrapper = factory({
      tabList: tabList
    })

    const activeTab = wrapper.find('.at-tabs__item--active')
    const tabs = wrapper.findAll('.at-tabs__item')
    expect(tabs[0].html()).toEqual(activeTab.html())

    await wrapper.setProps({ current: 2 })
    const activeTab1 = wrapper.find('.at-tabs__item--active')
    const tabs1 = wrapper.findAll('.at-tabs__item')
    expect(tabs1[2].html()).toEqual(activeTab1.html())
  })

  it('should render prop -- scroll', async () => {
    const wrapper = factory({
      tabList: tabList
    })

    expect(
      wrapper
        .find('.at-tabs--scroll')
        .exists()
    ).toBe(false)

    expect(
      wrapper
        .find('view.at-tabs__header')
        .exists()
    ).toBe(true)

    expect(
      wrapper
        .find('scroll-view.at-tabs__header')
        .exists()
    ).toBe(false)

    await wrapper.setProps({ scroll: true })
    expect(
      wrapper
        .find('.at-tabs--scroll')
        .exists()
    ).toBe(true)

    expect(
      wrapper
        .find('view.at-tabs__header')
        .exists()
    ).toBe(false)

    const scrollView = wrapper.find('scroll-view.at-tabs__header')
    expect(scrollView.exists()).toBe(true)
    expect(scrollView.element).toMatchSnapshot()
  })

  it('should render prop -- tabDirection', async () => {
    const wrapper = factory({
      tabList: tabList
    })

    expect(
      wrapper
        .find('.at-tabs--horizontal')
        .exists()
    ).toBe(true)
    expect(
      wrapper
        .find('.at-tabs--vertical')
        .exists()
    ).toBe(false)

    expect(
      wrapper
        .find('.at-tabs__underline')
        .attributes('style')
    ).toContain(`width: ${tabList.length * 100}%;`)

    expect(
      wrapper
        .find('.at-tabs__body')
        .attributes('style')
    ).toContain(`transform: translate3d(-0%, 0px, 0px);`)

    expect(wrapper.element).toMatchSnapshot()

    await wrapper.setProps({
      tabDirection: 'vertical',
      scroll: true
    })

    await wrapper.vm.$nextTick()

    expect(
      wrapper
        .find('.at-tabs--horizontal')
        .exists()
    ).toBe(false)
    expect(
      wrapper
        .find('.at-tabs--vertical')
        .exists()
    ).toBe(true)

    expect(
      wrapper
        .find('.at-tabs__underline')
        .attributes('style')
    ).toContain(`height: ${tabList.length * 100}%;`)

    expect(
      wrapper
        .find('.at-tabs__body')
        .attributes('style')
    ).toContain(`transform: translate3d(0px, -${0 * 100}%, 0px);`)

    expect(
      wrapper
        .find('.at-tabs__header')
        .attributes('scrolly')
    ).toBe('true')

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop -- animated', () => {
    const wrapper = factory({
      tabList: tabList,
      animated: false
    })

    expect(
      wrapper
        .find('.at-tabs__body')
        .attributes('style')
    ).toContain('transition: unset;')
  })

  it('should render prop -- height', () => {
    const wrapper = factory({
      tabList: tabList,
      tabDirection: 'vertical',
      height: '300px',
      // scroll: true,
    })

    expect(
      wrapper
        .find('.at-tabs__header')
        .element
    ).toMatchSnapshot()

    expect(
      wrapper
        .find('.at-tabs')
        .attributes('style')
    ).toContain('height: 300px;')

    // expect(
    //   wrapper
    //     .find('.at-tabs__header')
    //     .attributes('style')
    // ).toContain('height: 300px;')

    expect(
      wrapper
        .find('.at-tabs__body')
        .attributes('style')
    ).toContain('height: 300px;')
  })

  it('should render slot content', () => {
    const wrapper = factory(
      { tabList },
      { default: () => [h('view', { class: "test" }, "slot content")] }
    )

    expect(
      wrapper
        .find('.at-tabs__body > .test')
        .text()
    ).toBe('slot content')
  })
})

describe('AtTabs Behavior', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    jest.advanceTimersByTime(100)
    jest.runAllTimers()
  })

  it('should trigger onClick', async () => {
    const onClick = jest.fn()
    const wrapper = factory({
      tabList: tabList,
      onClick: onClick
    })

    const tabs = wrapper.findAll('.at-tabs__item')
    await tabs[1].trigger('tap')

    expect(onClick).toBeCalled()
  })

  it('should switch between tabs when swipeable is true', async () => {
    const current = ref(0)
    const handleClick = jest.fn((index) => {
      current.value = index
    })

    const wrapper = factory(
      {
        tabList,
        current,
        swipeable: true,
        onClick: handleClick
      },
      { default: () => [h('view', { class: "test" }, "slot content")] }
    )

    expect(
      wrapper
        .findAll('.at-tabs__item')[0]
        .classes()
    ).toContain('at-tabs__item--active')

    // swipe towards left
    let bodyEl = wrapper.find('.at-tabs__body')
    await triggerTouchEvents(
      bodyEl,
      {
        clientX: 372,
        clientY: 392,
        pageX: 372,
        pageY: 792,
      },
      {
        clientX: 103,
        clientY: 392,
        pageX: 103,
        pageY: 792,
      }
    )

    await wrapper.vm.$nextTick()
    expect(current.value).toBe(1)
    expect(
      wrapper
        .findAll('.at-tabs__item')[1]
        .classes()
    ).toContain('at-tabs__item--active')

    // swipe towards left
    bodyEl = wrapper.find('.at-tabs__body')
    await triggerTouchEvents(
      bodyEl,
      {
        clientX: 372,
        clientY: 392,
        pageX: 372,
        pageY: 792,
      },
      {
        clientX: 103,
        clientY: 392,
        pageX: 103,
        pageY: 792,
      }
    )

    expect(current.value).toBe(2)
    expect(
      wrapper
        .findAll('.at-tabs__item')[2]
        .classes()
    ).toContain('at-tabs__item--active')

    // swipe towards right
    bodyEl = wrapper.find('.at-tabs__body')
    await triggerTouchEvents(
      bodyEl,
      {
        clientX: 40,
        clientY: 392,
        pageX: 40,
        pageY: 792,
      },
      {
        clientX: 300,
        clientY: 392,
        pageX: 300,
        pageY: 792,
      }
    )

    expect(current.value).toBe(1)
    expect(
      wrapper
        .findAll('.at-tabs__item')[1]
        .classes()
    ).toContain('at-tabs__item--active')
  })

  it('should not be swipeable when tab direction is vertical', async () => {
    const current = ref(0)
    const handleClick = jest.fn((index) => {
      current.value = index
    })

    const wrapper = factory(
      {
        tabList,
        current,
        swipeable: true,
        tabDirection: "vertical",
        onClick: handleClick
      },
      { default: () => [h('view', { class: "test" }, "slot content")] }
    )

    // swipe towards left
    let bodyEl = wrapper.find('.at-tabs__body')
    await triggerTouchEvents(
      bodyEl,
      {
        clientX: 372,
        clientY: 392,
        pageX: 372,
        pageY: 792,
      },
      {
        clientX: 103,
        clientY: 392,
        pageX: 103,
        pageY: 792,
      }
    )

    expect(handleClick).not.toBeCalled()
    expect(current.value).not.toEqual(1)
    expect(current.value).toBe(0)
  })
})