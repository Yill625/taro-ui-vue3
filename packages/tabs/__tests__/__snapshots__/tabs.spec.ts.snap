// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtTabs should render default AtTabs 1`] = `
<view
  class="at-tabs at-tabs--horizontal at-tabs--WEB"
>
  <view
    class="at-tabs__header"
    id="__tests__2021"
  />
  <view
    class="at-tabs__body"
    style="transform: translate3d(-0%, 0px, 0px);"
  >
    <view
      class="at-tabs__underline"
      style="width: 0%;"
    />
    <!---->
  </view>
</view>
`;

exports[`AtTabs should render prop -- height 1`] = `
<view
  class="at-tabs__header"
  id="__tests__2021"
>
  <view
    class="at-tabs__item at-tabs__item--active"
    id="tab__tests__20210"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页1
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
  <view
    class="at-tabs__item"
    id="tab__tests__20211"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页2
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
  <view
    class="at-tabs__item"
    id="tab__tests__20212"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页3
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
</view>
`;

exports[`AtTabs should render prop -- scroll 1`] = `
<scroll-view
  class="at-tabs__header"
  id="__tests__2021"
  scrollintoview=""
  scrollwithanimation="true"
  scrollx="true"
  scrolly="false"
>
  <view
    class="at-tabs__item at-tabs__item--active"
    id="tab__tests__20210"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页1
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
  <view
    class="at-tabs__item"
    id="tab__tests__20211"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页2
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
  <view
    class="at-tabs__item"
    id="tab__tests__20212"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页3
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
</scroll-view>
`;

exports[`AtTabs should render prop -- tabDirection 1`] = `
<view
  class="at-tabs at-tabs--horizontal at-tabs--WEB"
>
  <view
    class="at-tabs__header"
    id="__tests__2021"
  >
    <view
      class="at-tabs__item at-tabs__item--active"
      id="tab__tests__20210"
    >
      <text
        style="white-space: nowrap;"
      >
        标签页1
      </text>
      <view
        class="at-tabs__item-underline"
      />
    </view>
    <view
      class="at-tabs__item"
      id="tab__tests__20211"
    >
      <text
        style="white-space: nowrap;"
      >
        标签页2
      </text>
      <view
        class="at-tabs__item-underline"
      />
    </view>
    <view
      class="at-tabs__item"
      id="tab__tests__20212"
    >
      <text
        style="white-space: nowrap;"
      >
        标签页3
      </text>
      <view
        class="at-tabs__item-underline"
      />
    </view>
  </view>
  <view
    class="at-tabs__body"
    style="transform: translate3d(-0%, 0px, 0px);"
  >
    <view
      class="at-tabs__underline"
      style="width: 300%;"
    />
    <!---->
  </view>
</view>
`;

exports[`AtTabs should render prop -- tabDirection 2`] = `
<view
  class="at-tabs at-tabs--scroll at-tabs--vertical at-tabs--WEB"
>
  <scroll-view
    class="at-tabs__header"
    id="__tests__2021"
    scrollintoview=""
    scrollwithanimation="true"
    scrollx="false"
    scrolly="true"
  >
    <view
      class="at-tabs__item at-tabs__item--active"
      id="tab__tests__20210"
    >
      <text
        style="white-space: nowrap;"
      >
        标签页1
      </text>
      <view
        class="at-tabs__item-underline"
      />
    </view>
    <view
      class="at-tabs__item"
      id="tab__tests__20211"
    >
      <text
        style="white-space: nowrap;"
      >
        标签页2
      </text>
      <view
        class="at-tabs__item-underline"
      />
    </view>
    <view
      class="at-tabs__item"
      id="tab__tests__20212"
    >
      <text
        style="white-space: nowrap;"
      >
        标签页3
      </text>
      <view
        class="at-tabs__item-underline"
      />
    </view>
  </scroll-view>
  <view
    class="at-tabs__body"
    style="transform: translate3d(0px, -0%, 0px);"
  >
    <view
      class="at-tabs__underline"
      style="width: 300%; height: 300%;"
    />
    <!---->
  </view>
</view>
`;

exports[`AtTabs should render prop -- tabList 1`] = `
<view
  class="at-tabs__header"
  id="__tests__2021"
>
  <view
    class="at-tabs__item at-tabs__item--active"
    id="tab__tests__20210"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页1
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
  <view
    class="at-tabs__item"
    id="tab__tests__20211"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页2
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
  <view
    class="at-tabs__item"
    id="tab__tests__20212"
  >
    <text
      style="white-space: nowrap;"
    >
      标签页3
    </text>
    <view
      class="at-tabs__item-underline"
    />
  </view>
</view>
`;
